Glitch attenuation, 3 and 4 point averaging

ctzsnooze opened the pull request:
Provides:

- 3 and 4 point feedforward smoothing options using LMA filtering, 
- glitch attenuation, 
- duplicate RC packet feed forward glitch minimisation,
- first order smoothing of feed forward steps; no need to use 2 point averaging by default

3 point averaging eliminates the 20hz periodicity seen in TBS CRSF Rx systems with OpenTx 2.2 and 2.3, making for a much smoother FF trace. It is best used with (probably should only be used with) locked 150hz mode. 3 point smoothing in 50hz mode will be smoother, but slower, than the default 2 point smoothing. CLI `set ff_interpolate_sp = AVERAGE3`

4 point averaging completely eliminates the cyclical 4-step periodicity in feed forward seen with R9 Rx when in 150hz mode with OpenTx 2.2 and 2.3. R9's cannot be locked in 150hz mode, so 4 point averaging will give smooth but very delayed feed forward signals when it goes to 50hz mode. This option is primarily intended when R9's are for smooth HD flying.  4 point averaging may also be useful for HD quads to reduce large FF steps. CLI `set ff_interpolate_sp = AVERAGE4`.

The new smoothing option `set ff_smooth_factor = 37` removes the need for the default two point LMA averaging by default.  Hence the new default is `set ff_interpolate_sp = ON`, which activates interpolated FF but without any averaging.

To  activate 2-point averaging, CLI `set ff_interpolate_sp = AVERAGE2`

Glitches are minimised by:

- interpolating forwards the previous feed forward and acceleration values, and not having an exaggerated setup later, whenever single duplicate packets result in a 'flat point' in the setpoint line.  This almost totally removes the glitching in FF that otherwise happens for single 'lost packets' in the RC signal.  If the preceding packet had a larger than usual step up (ie, an 'early' FrSky packet), it will not be interpolated forward.

- attenuating extreme boost spikes, regardless of cause, with a nonlinear attenuation method.  Does not attenuate return from a full rate flip.

Note: when using this with OpenTx, the ADC should (must) be off. (long press MENU, page to HARDWARE, uncheck ADC).

RC smoothing with BIQUAD of 50hz on input and PT1 at 100hz on derivative give lots of responsiveness and very smooth P and FF traces.  

For radios with regular packets and occasional duplicate data packets, (Futaba, some Spektrum and good FrSky X series Rx's), no averaging can work very well `set ff_interpolate_sp = ON`

Big THANK YOU to iCr4sh for setting up the proper LMA code for this.

mikeller commented on src/main/flight/interpolated_setpoint.c:
```diff
@@ -29,24 +29,43 @@
+        case FF_INTERPOLATE_AVG4:
+            j = 4;
+            break;
+    }
```
We can just use a `uint8_t` value with an appropriate input range (1 to 4), and avoid the switch statement.

ctzsnooze replied:
Thanks, much better!

mikeller commented on src/main/flight/interpolated_setpoint.c:
```diff
@@ -58,9 +77,36 @@ FAST_CODE_NOINLINE float interpolatedSpApply(int axis, bool newRcFrame, ffInterp
+        // Glitch reduction code for identical packets
+        if ((setpointSpeed == 0) && (fabsf(rawSetpoint) < 0.98f * ffMaxRate[axis])) {
+            // identical packets, not at full deflection
+            if ((holdCount[axis] < holdSteps) && (fabsf(rawSetpoint) > 2.0f) && (bigStep[axis] == 0)){
```
Missing whitespace. :space_invader:

ctzsnooze replied:
sorry can't figure where you mean - happy to fix, just not sure where

ctzsnooze replied:
ah, thanks Miguel - fixed

mikeller commented on src/main/flight/interpolated_setpoint.c:
```diff
@@ -58,9 +77,36 @@ FAST_CODE_NOINLINE float interpolatedSpApply(int axis, bool newRcFrame, ffInterp
+        const float holdSteps = 2.0f;
+
+        // Glitch reduction code for identical packets
+        if ((setpointSpeed == 0) && (fabsf(rawSetpoint) < 0.98f * ffMaxRate[axis])) {
```
Unneeded braces.

ctzsnooze replied:
thanks

ctzsnooze replied:
Ah OK it was the inner braces.  Removed.

mikeller commented on src/main/flight/interpolated_setpoint.c:
```diff
@@ -58,9 +77,36 @@ FAST_CODE_NOINLINE float interpolatedSpApply(int axis, bool newRcFrame, ffInterp
-        const float setpointSpeed = (rawSetpoint - prevRawSetpoint[axis]) * rxRate;
-        const float setpointAcceleration = (setpointSpeed - prevSetpointSpeed[axis]) * pidGetDT();
+        float setpointSpeed = (rawSetpoint - prevRawSetpoint[axis]) * rxRate;
+        float setpointAcceleration = (setpointSpeed - prevSetpointSpeed[axis]);
```
Unneeded braces.

ctzsnooze replied:
whoops - thank you

mikeller commented on src/main/flight/interpolated_setpoint.c:
```diff
@@ -58,9 +77,36 @@ FAST_CODE_NOINLINE float interpolatedSpApply(int axis, bool newRcFrame, ffInterp
+        // Glitch reduction code for identical packets
+        if ((setpointSpeed == 0) && (fabsf(rawSetpoint) < 0.98f * ffMaxRate[axis])) {
+            // identical packets, not at full deflection
+            if ((holdCount[axis] < holdSteps) && (fabsf(rawSetpoint) > 2.0f) && (bigStep[axis] == 0)){
```
Unneeded braces.

ctzsnooze replied:
sorry

ctzsnooze replied:
Removed inner braces.

mikeller commented on src/main/flight/interpolated_setpoint.c:
```diff
@@ -58,9 +77,36 @@ FAST_CODE_NOINLINE float interpolatedSpApply(int axis, bool newRcFrame, ffInterp
+        // Glitch reduction code for identical packets
+        if ((setpointSpeed == 0) && (fabsf(rawSetpoint) < 0.98f * ffMaxRate[axis])) {
+            // identical packets, not at full deflection
+            if ((holdCount[axis] < holdSteps) && (fabsf(rawSetpoint) > 2.0f) && (bigStep[axis] == 0)){
```
`bigStep[axis] == 0` is comparing a bool to an int.

If you call the variable `useBigStep` instead, `if (useBigStep[axis]) {` will become much easier to read and understand.

ledvinap replied:
@mikeller : the variable is 'there was big step in previous iteration', so bigStep is OK.
but `!bigStep[axis]` is obviously cleaner

ctzsnooze replied:
thanks - one day maybe I'll know .0001% of what you guys do :-)

mikeller commented on src/main/flight/interpolated_setpoint.c:
```diff
@@ -58,9 +77,36 @@ FAST_CODE_NOINLINE float interpolatedSpApply(int axis, bool newRcFrame, ffInterp
+            // we're moving!
+            holdCount[axis] = 1.0f;
+            if (fabsf(setpointAcceleration - prevAcceleration[axis]) > 1000.0f) {
+                bigStep[axis] = 1;
```
Should be `true` / `false`.

ctzsnooze replied:
thanks, done

ctzsnooze commented on src/main/flight/interpolated_setpoint.c:
```diff
@@ -29,24 +29,43 @@
 static float setpointDeltaImpl[XYZ_AXIS_COUNT];
-static float prevSetpointDeltaImpl[XYZ_AXIS_COUNT];
 static float setpointDelta[XYZ_AXIS_COUNT];
+static float holdCount[XYZ_AXIS_COUNT];
```
thanks, I didn't know if fast code worked better if everything was floats, sorry

ledvinap replied:
It usually does in calculations (no int->float conversion). But holdCount  is simple counter, compared against integers

ctzsnooze commented on src/main/flight/interpolated_setpoint.c:
```diff
@@ -29,24 +29,43 @@
     for (int i = 0; i < XYZ_AXIS_COUNT; i++) {
         ffMaxRate[i] = applyCurve(i, 1.0f);
         ffMaxRateLimit[i] = ffMaxRate[i] * ffMaxRateScale;
+        laggedMovingAverageInit(&setpointDeltaAvg[i].filter, j, (float *)&setpointDeltaAvg[i].buf);
```
thanks, made that change... very grateful for your help

ctzsnooze commented:
Thanks @mikeller and @ledvinap for your advice, have implemented what I can.
As Rx signals improve, much of this won't be needed.  But now that we are leveraging feed forward more strongly than ever, and for considerable benefit, these little hacks are very useful.

ctzsnooze commented:
@mikeller @etracer65 @joelucid ... two queries for discussion:

1.  The old default was to average 2 feed forward steps at all times.  This was to avoid large glitches from duplicate incoming RC data steps.  The glitch removal code cleans these up much better, and averaging two introduces a lot of delay, especially in 50hz modes.  So I would like to change the default to just `ON`.

2.  Perhaps we could improve the user-friendliness with the following changes to naming of the the CLI entries:

- change `ff_interpolate_sp` to `ff_interpolation_mode`
- change the `OFF, ON, AVERAGE2, AVERAGE3, AVERAGE4' labels to `OFF, ON, AVG2, TBS, R9`

The intent is to make it easier for the relevant users to select the correct mode for their Rx.

mikeller commented:
@ctzsnooze: Re 2., remember, CLI is not where user friendliness is key - this is for early adopters, and to prove its value and that it's done in the best way. Once this is done we'll add support for it to configurator, with more verbose setting names and helptexts at all. I'd just dump `lookupTableInterpolatedSetpoint` altogether and use the numerical value directly - adding text labels for different numerical sample sizes seems to be moot.

joelucid commented:
I agree with @mikeller that in this case numbers might work. Saves a couple of bytes of flash, too. I’m all for reducing latency - if the glitch avoidance code works well enough let’s default to single point mode.

etracer65 commented:
> change the OFF, ON, AVERAGE2, AVERAGE3, AVERAGE4 labels to OFF, ON, AVG2, TBS, R9

I don't think including a manufacturer name in the settings options is a good idea. What happens when TBS or Frysky come out with a firmware update that fixes one of the problems you're trying to work around? It would be even more confusing to tell a R9 user to use the "TBS" setting or tell a CRSF user to not use the "TBS" setting as examples.

netzwergracer commented:
Works great with R9M and 4 point averaging!
![Unbenannt](https://user-images.githubusercontent.com/28099853/67699976-5dcec080-f9ad-11e9-8395-f2356dc3140b.PNG)

ctzsnooze commented:
Thanks for the feedback.
If we dumped `lookupTableInterpolatedSetpoint` then I suppose `ff_interpolated_sp` could be re-named just `ff_interpolated` where 0=off, 1 = on without averaging, 2 = on with a two point moving average, 3 = on with a three point moving average, and so on.
Hmm.
I kind of like the longer 'text' version - it is a bit more intuitive - would it be OK to keep that, but rename it `ff_interpolate_mode`?

mikeller commented:
@ctzsnooze: `ff_interpolate_mode` sounds good. And since we're already changing from lookup to integer the change in name will not matter with respect to restoring from `diff`.

ctzsnooze commented:
On reflection, when ff_interpolated_sp is off, feed forward still exists, but is calculated per loop as the simple derivative of setpoint.  Perhaps confusingly, setpoint itself can have a linear interpolation applied as an alternative to filter based smoothing.
This new method gets raw setpoint and generates steps; it allows using 'boost' or 'acceleration' and feed forward limiting when sticks reach max trace, etc, none of which is possible with the simpler feed forward method.
In that sense the options here are, conceptually, the 'old' way, the 'new way' without averaging, the 'new way with 2 point averaging' etc.
Maybe something along these lines:
ff_mode: classic, new, average2, average3, average4
ff_mode: derivative, interpolated, average2, average3, average4
??

mikeller commented:
@ctzsnooze: Again, this is not where user friendliness is key - these should be technically accurate descriptions of **what it is** and not of what it is meant to be used for. And in this case I think the 'number of samples that we are averaging over' is an accurate description.
Adding the nice UI and helpful explanations should be done in configurator when we get to that stage.

McGiverGim commented on src/main/flight/interpolated_setpoint.c:
```diff
@@ -58,9 +66,36 @@ FAST_CODE_NOINLINE float interpolatedSpApply(int axis, bool newRcFrame, ffInterp
+        // Glitch reduction code for identical packets
+        if (setpointSpeed == 0 && fabsf(rawSetpoint) < 0.98f * ffMaxRate[axis]) {
+            // identical packets, not at full deflection
+            if (holdCount[axis] < holdSteps && fabsf(rawSetpoint) > 2.0f && !bigStep[axis]){
```
In the first review, the `Missing whitespace. 👾` message was because you need to add a space between the last `)` and the `{`.

ctzsnooze replied:
thanks McGiverGim - fixed!
@mikeller - OK no problem to stick with off, on, average2, average3 etc.

mikeller replied:
@ctzsnooze: Ok, then we should change the lookup table elements to be indicative of what they actually represent: `OFF`, `ON`, `AVERAGED_2`, `AVERAGED_3`, ...

ctzsnooze replied:
I will do this prior to squashing and rebasing

ctzsnooze replied:
Averaging types renamed `AVERAGED_2`, `AVERAGED_3` etc..

ctzsnooze commented:
Added a significant improvement to overall smoothing of the FF steps and a small improvement to the glitch removal code.

FF steps from this interpolated method at present can vary up and down with timing irregularities anywhere in the rc chain.  Boost makes these worse.  The glitch removal code is active only when a packet is identical to the previous packet, and doesn't reduce these more generalised irregularities.  The clipping code identifies and suppresses very large FF steps, but doesn't smooth these kinds of smaller scale jitters at all.

rc_derivative smoothing sort of rounds the leading edge of a step off, but doesn't smooth the height of the steps.  When set relatively low, a jittery FF trace becomes a series of sharp triangular ups and downs with significant rc derivative smoothing.

2 point averaging is the current default method for attenuating these jitters by 'smoothing' up and down variations.  But it is quite problematic because it causes delayed artefacts after single steps up or down.  It is useful if there is regular two-step periodicity, just as three step moving averaging is good for regular three-step periodicity, but for smoothing random jitter, a moving average is not the optimal method.

I did quite a lot of modelling and confirmed that a PT1 style filtering method works much better than 2 point averaging.  I do not use a PT1 filter for this, but enter the k value directly via the CLI as a 'smoothing' value.  The default is a one time constant value.  When the smoothing is set to the default of 37, each FF step will change by 63% of what it otherwise would.  This introduces a fair bit of smoothing without late consequences of significance.  It is approximately equivalent to a lowpass filter set at 2 times the packet interval.  So for 9ms steps, the default approximates an 18hz PT1, and for 7ms steps, about 23hz.  This provides approximately the same 'delay' as the current 2 point averaging, but with a smoother and cleaner outcome.

By lowering the smoothing factor, eg to say 18, we would get the equivalent of about a 40hz PT1 being applied to the steps themselves.

A zero smoothing factor has no effect.

A higher smoothing will result in even smoother, but more delayed, feed forward.  Since the whole goal of feed forward is to provide quick to immediate PID responses to stick movements, it becomes pointless to smooth feed forward a great deal.   I limited the maximum smoothing factor to 75, meaning that successive FF steps will only be 25% of what they otherwise would be.  This may be useful for cinematic quads, however, where FF jitter of even small amounts can cause wobble in the HD image.

I also improved the glitch identification code a little.  The code runs when two identical packets arrive in succession.  Most commonly, the preceding packets are 'normal', then there are two identical packets followed by a double height up-step when setpoint returns to the normal stick trajectory.  For these kind of issues, the code projects forward the feed forward components into the 'gap', and now halves the acceleration and feed forward quantum for the next big step back up, resulting mostly in a perfectly smooth FF trace despite the 'lost' data.  

FrSky radios often 'get a head of themselves' and transmit a double-height step up followed by an identical packet.  Without glitch code, these generate a large FF and boost spike, followed by a zero value for FF and a strongly negative boost value.  If we were to interpolate forward the large upwards spike, we would get a double-width FF and boost spike, causing gyro to overshoot significantly.  The `PREV_BIG_STEP` value is a threshold which attempts to identify these situations and handle them differently from the more common 'lost packet' situations.  The idea is that if the preceding acceleration value, before the unchanged data packet, was 'unreasonably large', then most likely was is one of these 'early FrSky' packets; in this case, we do not interpolate it forward, and we allow a negative correction.  These kinds of glitches are effectively not attenuated by the de-glitching code, since I cannot figure any way to identify them in the time frame needed to make a correction.  They are only attenuated by they clipping code, and even then we can't clip that hard otherwise 'real' fast stick inputs would lose too much FF.

The value for `PREV_BIG_STEP` is arbitrary and has been empirically determined from my testing; it seems to be a reasonable compromise.  Nearly all 'lost' packets get handled appropriately, and differently from the 'early' packets, with this method.  The debug shows steps that are identified as big steps, and the counter goes up for 'lost' packets and down for 'early' packets as determined by the previous step size threshold, for debugging purposes.

I am not happy with the final complexity of the code, but the performance is now very, very good indeed.

ctzsnooze commented:
This graphic shows how 2 point LMA introduces an artefact one step after after any strong change, whereas a PT1 type approach doesn't.
The un-filtered FF (plain feed forward plus boost) is in green, lowpass filtered is red, 2 point moving average is orange.  Default smoothing factor of 37 results in similar smoothing.  2 point LMA causes a false step seen at steps 11 and 16.

![Screen Shot 2019-11-06 at 17 19 18](https://user-images.githubusercontent.com/11737748/68552985-60331f00-0471-11ea-9f5b-7c8c54bfe65d.jpg)

Here is the same input with smoothing at 20, clearly the spike is less attenuated, this would be alright if the feed forward signal was smooth and relatively jitter free...
![Screen Shot 2019-11-06 at 17 19 40](https://user-images.githubusercontent.com/11737748/68552997-7e008400-0471-11ea-8fb5-59fa0d828080.jpg)

ctzsnooze commented:
This graphic shows an episode of regular up/down stepping.  The 2 point moving average (orange) eliminates it completely, but the cost is the delayed false glitch upwards at 17 after downwards at 16.  The smoothing approach attenuates the up/down stepping by almost half, but does not eliminate it.  However there is no late glitch at all.

We can see also that at at the sharp reversal, the 2 point averaging markedly delays the peak FF response by an entire packet, whereas the smoothing method works faster.
![Screen Shot 2019-11-06 at 17 58 14](https://user-images.githubusercontent.com/11737748/68553048-ecdddd00-0471-11ea-8087-79a8683f8670.jpg)

ctzsnooze commented:
The end result can be very, very good.. this bit of a log has quite a lot of RC glitches of different kinds, but they have almost no impact on performance, which is excellent, with gyro tracking setpoint with uncanny accuracy...
![Screen Shot 2019-11-09 at 15 32 05](https://user-images.githubusercontent.com/11737748/68553162-df752280-0472-11ea-9559-8a709f0bf0eb.jpg)

ctzsnooze commented:
This is an example of the up/down glitch counter debug in purple at the top.  It changes state when a packet is the same as the previous packet, going up if the preceding step up wasn't huge, and down if greater than the big step threshold.  This is a typical FrSky trace.  Sometimes the step threshold isn't perfect but its usually correct.

![Screen Shot 2019-11-09 at 12 06 52](https://user-images.githubusercontent.com/11737748/68553222-63c7a580-0473-11ea-9ef1-60f3e95e3ac3.jpg)

bexamous commented on src/main/flight/interpolated_setpoint.c:
```diff
@@ -58,9 +69,42 @@ FAST_CODE_NOINLINE float interpolatedSpApply(int axis, bool newRcFrame, ffInterp
+                holdCount[axis] += 1.0f;
+            // hit timeout, or at centre and not moving, or prev big step, lock speed and acceleration to zero
+            } else {
+                holdCount[axis] -= 1.0f;
```
Is this correct?  holdSteps is currently 2, which allows us to 'replace' one repeated frame with a 'guess'...

But what about frames going 1,2,3,4,4,4,4

The first 4 is good, holdCount=1.
The 2nd 4 is repeated frame, holdCount is 1 and 1<2 so we use previous speed + acceleration and set holdCount=2.
The 3rd 4 is again repeated..  holdCount=2 so we hit this else block and line 86 sets holdCount-=1, so its 1 again.
The 4th 4 is again repeated.. holdCount=1 again, so now we can get back into line 80 if block?  That doesn't seem like we want that?

Is line 86 needed?  If deleted I think things might be fine, or maybe its even better to do holdCount[axis] = holdSteps, so if somehow we get in this else block we 'disable' any holding going forward until we get a non-repeated frame?

bexamous replied:
Oh I remember other issue with doing holdCount[axis] -= 1...  this is an unsigned int.  What happens if you start getting repeated frames but sticks are centered?  Won't you will keep getting into this else block over and over again long as sticks stay centered right?  And each time you get into it holdCount-=1... only will have to repeat a couple frames before holdCount=0 and you do holdCount-=1 and it becomes INT_MAX.

ctzsnooze replied:
It could be that this can be improved.  We need a way to know, when the sticks move again, what the status was before they moved. 
In the example you gave, 1,2,3,4,4,4,4,5,5 (adding 5's when sticks start), we only want normal acceleration and boost.

ctzsnooze replied:
Maybe line 86 needs to be `holdCount[axis] = 3.0f;`
Then after the 3,4 transition, the first 4,4 will set holdCount to 2, and do the right things, then the second 4,4 will set holdCount to 3, and from then on while setpoint speed is zero, holdCount will stay 3. 
When we start moving, only if holdCount is 2 (meaning one identical packet, ie the first 4,4) will we reduce the FF and boost.  If holdCount was 3, ie flat, line 98 will set it back to 1 and be in normal detection state.
I think this will solve the problem?

ctzsnooze replied:
set to above solution, tested, works well.  thanks

bexamous commented on src/main/flight/interpolated_setpoint.c:
```diff
@@ -58,9 +69,42 @@ FAST_CODE_NOINLINE float interpolatedSpApply(int axis, bool newRcFrame, ffInterp
+            } else {
+                holdCount[axis] -= 1.0f;
+                setpointSpeed = 0.0f;
+                setpointAcceleration = 0.0f;
```
I don't undertand this?  To even get here setpointSpeed is already 0, so I think that's unneeded.  But why do we set acceleration to 0?  Alternative would be leaving it set to whatever line 73 calculated it to be?  Is 0 better?

ctzsnooze replied:
Yes, setting setpoint speed to 0 is not needed in 87, thanks. 
Acceleration depends on prevSetpointSpeed[axis]; this won't be zero if projected forward like  when holdCount was 2.  It does need to be forced to zero.

bexamous commented on src/main/flight/interpolated_setpoint.c:
```diff
@@ -58,9 +69,42 @@ FAST_CODE_NOINLINE float interpolatedSpApply(int axis, bool newRcFrame, ffInterp
+            }
+        } else {
+            // we're moving!
+            if (holdCount[axis] == 2.0f) {
```
Currently this makes sense, but what happens if you set holdSteps=3?  And you hold 2 frames and then get here with holdCount[axis]=3?

Maybe something like:

```
if (holdCount[axis] > 1) {
    setpointSpeed /= holdCount[axis];
    setpointAcceleration /= holdCount[axis];
}

So then if we hold for 2 frames, holdCount[axis] is then 3, and we would expect next 'good' package to be 3 times what it should be so we device by 3.
```

ctzsnooze replied:
I originally did code it to potentially handle holds for more than one step, but to keep it simple enough for now, it only manages one single missed packet.  That was complicated enough.  The problem what to do after re-starting after more than one identical packet is knowing whether the train of identical packets was real or not.  If the user held the sticks still, or wasn't moving much, or had ADC on, then we get runs of identical packets and not necessarily a big step up.  If we attenuate that step up after say 10 identical packets, and actually it is a 'normal' step up like with ADC on, that would be bad.  
I was hoping to sort this code out as is, then extend it possibly for longer gaps.

ctzsnooze replied:
I won't do this for now, but it is something we could consider later, it is a bit tricky to test.

ctzsnooze commented on src/main/flight/interpolated_setpoint.c:
```diff
@@ -58,9 +69,42 @@ FAST_CODE_NOINLINE float interpolatedSpApply(int axis, bool newRcFrame, ffInterp
-        const float setpointAcceleration = (setpointSpeed - prevSetpointSpeed[axis]) * pidGetDT();
+        float setpointSpeed = (rawSetpoint - prevRawSetpoint[axis]) * rxRate;
+        float setpointAcceleration = setpointSpeed - prevSetpointSpeed[axis];
+        const float holdSteps = 2.0f;
```
Yes they could be integers, for sure.  I am not sure however if for code efficiency it is better in this FAST block that they are floats.  Happy to change if it would be better.  @etracer65 you thoughts?

mikeller replied:
@ctzsnooze: No, ints will be calculated faster.

ctzsnooze replied:
Changed all to uint8_t, thanks

ctzsnooze commented on src/main/flight/interpolated_setpoint.c:
```diff
@@ -58,9 +69,42 @@ FAST_CODE_NOINLINE float interpolatedSpApply(int axis, bool newRcFrame, ffInterp
+                // holding the entire previous speed is best for missed packets, but bad for early packets
+                setpointSpeed = prevSetpointSpeed[axis] + prevAcceleration[axis];
+                setpointAcceleration = prevAcceleration[axis];
+                holdCount[axis] += 1.0f;
```
yes, see above, happy to change this

ctzsnooze replied:
done

ctzsnooze commented on src/main/flight/interpolated_setpoint.c:
```diff
@@ -58,9 +69,42 @@ FAST_CODE_NOINLINE float interpolatedSpApply(int axis, bool newRcFrame, ffInterp
+                setpointSpeed = prevSetpointSpeed[axis] + prevAcceleration[axis];
+                setpointAcceleration = prevAcceleration[axis];
+                holdCount[axis] += 1.0f;
+            // hit timeout, or at centre and not moving, or prev big step, lock speed and acceleration to zero
```
ok

ctzsnooze replied:
done

ctzsnooze commented:
Thanks @bexamous @mikeller for those clarifications... 
Here is a small bench log with the counter indicating:
- 1 = normal, 
- 2 = a single identical packet without a preceding large step up, and 
- 3 = either a steady stick, or a single identical packet with a large preceding step up, but not at max.  
First and image showing the counter of 3 working correctly with clean signal:
![Screen Shot 2019-11-18 at 12 44 56](https://user-images.githubusercontent.com/11737748/69018681-950b1d00-0a01-11ea-8a62-ad693ff6f132.jpg)
Now an image with all types of events in an R-Xsr log; they seem to be handled appropriately:
![Screen Shot 2019-11-18 at 12 33 37](https://user-images.githubusercontent.com/11737748/69018710-afdd9180-0a01-11ea-9644-42946774bee1.jpg)

ctzsnooze commented:
Here is a log of a real flight with all kinds of Rx issues but the code doing OK.  This is the version where holdCount goes to 3.
[Log_HoldCount_3.BBL.zip](https://github.com/betaflight/betaflight/files/3856506/Log_HoldCount_3.BBL.zip)

ctzsnooze commented:
@bexamous I realised that now we have a specific holdCount value for early steps, it can be used to block the excess boost factor when we restart after one of these.  
On the left there are two early packets followed by identical packets.
Boost jumps twice, first at the early packet, then drops to zero, and then causes another push in the same direction when the next valid packet comes on.  However the end result is typically overshoot.
Now I force boost to zero after such an event, the image shows no additional boost on the next step, reducing, to some extent, the overshoot associated with these early packets.
The downside is that it will reduce the boost on the very start of a clean new stick movement from centre, however this is unlikely to be much of an issue since most valid human accelerations of sticks from still position are not instantaneous.
![Screen Shot 2019-11-18 at 13 45 24](https://user-images.githubusercontent.com/11737748/69020957-5f1e6680-0a0a-11ea-81fb-42f962ead9b2.jpg)

ctzsnooze commented:
@mikeller, @ledvinap, @etracer65 
Thanks for your feedback... I've implemented all suggestions received to date, and rebased and squashed.  
There has been a lot of testing, and I am confident it does what is needed.
I would be grateful for suggestions for code efficiency etc.
Default is to not use 2 point averaging but instead the PT1 style step smoothing.
Averaging likely won't be needed at all once OpenTx sync to external modules is implemented, but this will take a while and for now the averaging is useful.
If you're agreeable I think this is ready to be merged.

mikeller commented on src/main/flight/interpolated_setpoint.c:
```diff
@@ -28,25 +28,36 @@
 static float setpointDelta[XYZ_AXIS_COUNT];
+static uint8_t holdCount[XYZ_AXIS_COUNT];
 
+typedef struct laggedMovingAverageCombined_u {
```
As this is not a union but a struct, the name should be `laggedMovingAverageCombined_s` and not `laggedMovingAverageCombined_u`.

ctzsnooze replied:
Thanks!  Fixed.
